File: generated-code/php/codecraft/FileReadWrite/Codegame/MessageGameModel.php
<?php

namespace Codegame {
    
    require_once 'Codegame/ClientMessage.php';
    require_once 'Codegame/ServerMessage.php';

    /**
     * Client or server message
     */
    abstract class MessageGameModel
    {
        /**
         * Write MessageGameModel to output stream
         */
        abstract function writeTo($stream);

        /**
         * Read MessageGameModel from input stream
         */
        static function readFrom($stream)
        {
            $tag = $stream->readInt32();
            if ($tag == \Codegame\MessageGameModel\Client::TAG) {
                return \Codegame\MessageGameModel\Client::readFrom($stream);
            }
            if ($tag == \Codegame\MessageGameModel\Server::TAG) {
                return \Codegame\MessageGameModel\Server::readFrom($stream);
            }
            throw new Exception('Unexpected tag value');
        }
    }
}

namespace Codegame\MessageGameModel {
    /**
     * Client message
     */
    class Client extends \Codegame\MessageGameModel
    {
        const TAG = 0;
    
        /**
         * Message
         */
        public $message;
    
        function __construct($message)
        {
            $this->message = $message;
        }
    
        /**
         * Read Client from input stream
         */
        public static function readFrom($stream)
        {
            $message = \Codegame\ClientMessage::readFrom($stream);
            return new Client($message);
        }
        
        /**
         * Write Client to output stream
         */
        public function writeTo($stream)
        {
            $stream->writeInt32(Client::TAG);
            $this->message->writeTo($stream);
        }
    }

    /**
     * Server message
     */
    class Server extends \Codegame\MessageGameModel
    {
        const TAG = 1;
    
        /**
         * Message
         */
        public $message;
    
        function __construct($message)
        {
            $this->message = $message;
        }
    
        /**
         * Read Server from input stream
         */
        public static function readFrom($stream)
        {
            $message = \Codegame\ServerMessage::readFrom($stream);
            return new Server($message);
        }
        
        /**
         * Write Server to output stream
         */
        public function writeTo($stream)
        {
            $stream->writeInt32(Server::TAG);
            $this->message->writeTo($stream);
        }
    }
}

File: generated-code/cpp/codecraft/FileReadWrite/codegame/MessageGameModel.cpp
#include "MessageGameModel.hpp"
#include <stdexcept>

namespace codegame {

MessageGameModel::Client::Client(std::shared_ptr<codegame::ClientMessage> message) : message(message) { }

// Read Client from input stream
MessageGameModel::Client MessageGameModel::Client::readFrom(InputStream& stream) {
    std::shared_ptr<codegame::ClientMessage> message = codegame::ClientMessage::readFrom(stream);
    return MessageGameModel::Client(message);
}

// Write Client to output stream
void MessageGameModel::Client::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    message->writeTo(stream);
}

// Get string representation of Client
std::string MessageGameModel::Client::toString() const {
    std::stringstream ss;
    ss << "MessageGameModel::Client { ";
    ss << "message: ";
    ss << message->toString();
    ss << " }";
    return ss.str();
}

MessageGameModel::Server::Server(std::shared_ptr<codegame::ServerMessage> message) : message(message) { }

// Read Server from input stream
MessageGameModel::Server MessageGameModel::Server::readFrom(InputStream& stream) {
    std::shared_ptr<codegame::ServerMessage> message = codegame::ServerMessage::readFrom(stream);
    return MessageGameModel::Server(message);
}

// Write Server to output stream
void MessageGameModel::Server::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    message->writeTo(stream);
}

// Get string representation of Server
std::string MessageGameModel::Server::toString() const {
    std::stringstream ss;
    ss << "MessageGameModel::Server { ";
    ss << "message: ";
    ss << message->toString();
    ss << " }";
    return ss.str();
}

// Read MessageGameModel from input stream
std::shared_ptr<MessageGameModel> MessageGameModel::readFrom(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return std::shared_ptr<MessageGameModel::Client>(new MessageGameModel::Client(MessageGameModel::Client::readFrom(stream)));
    case 1:
        return std::shared_ptr<MessageGameModel::Server>(new MessageGameModel::Server(MessageGameModel::Server::readFrom(stream)));
    default:
        throw std::runtime_error("Unexpected tag value");
    }
}

}